apollo-router/src/apollo_telemetry.rs

```diff
@@ -0,0 +1,553 @@
+        // Convert the configuration data into a reportable form
+        let graph: ReporterGraph = self.graph.as_ref().unwrap().into();
+
+        // Report our consolidated statistics
+        for ((client_name, client_version, key), dh) in dh_map.drain() {
+            tracing::debug!("reporting entries: {}", dh.entries);
+            let stats = ContextualizedStats {
+                context: Some(StatsContext {
+                    client_name,
+                    client_version,
+                }),
+                query_latency_stats: Some(QueryLatencyStats {
+                    latency_count: dh.buckets,
+                    request_count: dh.entries,
+                    ..Default::default()
+                }),
+                ..Default::default()
+            };
+
+            let msg = reporter
+                .submit_stats(graph.clone(), key, stats)
+                .await
+                .map_err::<TraceError, _>(|e| e.to_string().into())?
+                .into_inner()
+                .message;
+            tracing::trace!("server response: {}", msg);
+        }
+
+        Ok(())
+    }
+}
+
+// Taken from TS implementation
+static GRAPHQL_PARSE_FAILURE: &str = "## GraphQLParseFailure\n";
+static GRAPHQL_VALIDATION_FAILURE: &str = "## GraphQLValidationFailure\n";
+static GRAPHQL_UNKNOWN_OPERATION_NAME: &str = "## GraphQLUnknownOperationName\n";
+
+fn normalize(op: Option<&opentelemetry::Value>, query: &str) -> String {
+    // If we don't have an operation name, we can't do anything useful
+    // with this query. Just return the appropriate error.
+    let op_name: String = match op {
+        Some(v) => v.as_str().into_owned(),
+        None => {
+            tracing::warn!("Could not identify operation name: {}", query);
+            return GRAPHQL_UNKNOWN_OPERATION_NAME.to_string();
+        }
+    };
+
+    let parser = Parser::new(query);
+    // compress *before* parsing to modify whitespaces/comments
+    let ast = parser.compress().parse();
+    tracing::debug!("ast:\n {:?}", ast);
+    // If we can't parse the query, we definitely can't normalize it, so
+    // just return the appropriate error.
+    if ast.errors().len() > 0 {
+        tracing::warn!("Could not parse query: {}", query);
+        return GRAPHQL_PARSE_FAILURE.to_string();
+    }
+    let doc = ast.document();
+    tracing::debug!("{}", doc.format());
+    tracing::trace!("looking for operation: {}", op_name);
+    let mut required_definitions: Vec<_> = doc
+        .definitions()
+        .into_iter()
+        .filter(|x| {
+            if let ast::Definition::OperationDefinition(op_def) = x {
+                return match op_def.name() {
+                    Some(v) => v.text() == op_name,
+                    None => op_name == "-",
+                };
+            }
+            false
+        })
+        .collect();
+    tracing::debug!("required definitions: {:?}", required_definitions);
+    if required_definitions.len() != 1 {
```

Comment:
shouldn't that be `required_definitions.len() == 0` ?

Reply:
I think it can happen that len() > 1. I'm not 100% sure to be honest, but I think it would happen if we had unnamed queries that matched multiple times. Ultimately, this code is going to be re-visited when normalization is more complete, so I think it's fine as is.

Reply:
ok